lib/api/confighandler.go

```diff
@@ -0,0 +1,336 @@
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package api
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/julienschmidt/httprouter"
+	"golang.org/x/crypto/bcrypt"
+
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/protocol"
+	"github.com/syncthing/syncthing/lib/sync"
+)
+
+type configMuxBuilder struct {
+	*httprouter.Router
+	id  protocol.DeviceID
+	cfg config.Wrapper
+	mut sync.Mutex
+}
+
+func (c *configMuxBuilder) registerConfig(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, c.cfg.RawCopy())
+	})
+	c.PUT(path, func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.adjustConfig(w, r)
+	})
+}
+
+func (c *configMuxBuilder) registerConfigDeprecated(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, c.cfg.RawCopy())
+	})
+	c.POST(path, func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.adjustConfig(w, r)
+	})
+}
+
+func (c *configMuxBuilder) registerConfigInsync(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, map[string]bool{"configInSync": !c.cfg.RequiresRestart()})
+	})
+}
+
+func (c *configMuxBuilder) registerFolders(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, c.cfg.FolderList())
+	})
+	c.PUT(path, func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+		var waiter config.Waiter
+		folders := make([]config.FolderConfiguration, 0)
+		err := unmarshalTo(r.Body, &folders)
+		if err == nil {
+			waiter, err = c.cfg.SetFolders(folders)
+		}
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		c.finish(w, waiter)
+	})
+}
+
+func (c *configMuxBuilder) registerDevices(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, c.cfg.DeviceList())
+	})
+	c.PUT(path, func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+		var waiter config.Waiter
+		devices := make([]config.DeviceConfiguration, 0)
+		err := unmarshalTo(r.Body, &devices)
+		if err == nil {
+			waiter, err = c.cfg.SetDevices(devices)
+		}
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		c.finish(w, waiter)
+	})
+}
+
+func (c *configMuxBuilder) registerFolder(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, p httprouter.Params) {
+		folder, ok := c.cfg.Folder(p.ByName("id"))
+		if !ok {
+			http.Error(w, "No folder with given ID", http.StatusNotFound)
+			return
+		}
+		sendJSON(w, folder)
+	})
+	c.PUT(path, func(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
+		c.adjustFolder(w, r, config.FolderConfiguration{})
+	})
+
```

Comment:
Spacing inconsistency OCD triggered. I think we should have a space between all the method registrations...